plugin/forward/persistent.go

```diff
@@ -129,7 +150,18 @@ func (t *transport) Dial(proto string) (*dns.Conn, bool, error) {
 func (t *transport) Yield(c *dns.Conn) { t.yield <- c }
 
 // Stop stops the transport's connection manager.
-func (t *transport) Stop() { close(t.stop) }
+func (t *transport) Stop() {
+	t.signal <- true
+	time.Sleep(100 * time.Millisecond)
+	close(t.stop)
```

Comment:
Lets assume `Stop()` is called when we are in the middle of `Dial()`. After successful `Dial()` we normally send the query, read response and yield the connection back to cache. Reading response may take more time than 100ms, so connManager() may already get stopped and `Yield()` will hang.
Perhaps, the code `if t.Signal() {` should be added to `Yield()` as you did it for `Dial()`